### DISCUSSION TITLE
Switch between multiple sets of bindings

### DISCUSSION THREAD

**Discussion by jacobtomlinson** [Textualize/textual#1306] [Q&A] [answered]
It would be great to support binding modes in my textual apps. Similar to command mode and edit mode in vim I'd like to be able to switch out different sets of bindings.

I see I can have different bindings associated with different screens but I'd like to be able to change between binding sets within one screen.

I tried updating the `BINDINGS` constant within an action with a new set of bindings but it didn't seem to have any effect (and modifying a constant feels icky).

**Comment by davep** [comment]
I what sort of situations are you finding you want to swap bindings within a screen? What sort of context causes the wish to swap around? I ask only because, last week, I sort of found myself wanting to do the same but I realise that what I really wanted to be doing was [to have relevant bindings at the widget level](https://textual.textualize.io/blog/2022/11/26/on-dog-food-the-original-metaverse-and-not-being-bored/#dont-put-all-your-bindings-in-one-place).

**Comment by jacobtomlinson** [comment]
Oh interesting I didn't realise you could add bindings to widgets! So would you have different bindings when different widgets are in focus?

I have about 8 widgets on my screen, but I only want two of them to be focusable, and one or other should always be in focus. Right now to achieve this I'm not using focus and doing things manually, but I can definitely change that.

**Comment by davep** [comment]
> So would you have different bindings when different widgets are in focus?

Yup, that's what I did with [this app](https://github.com/davep/unbored). Mostly I don't want any bindings other than a couple of very general ones, but when focus is within one of the activity displays, [I have a whole bunch of bindings on that widget](https://github.com/davep/unbored/blob/452afa472ece34c69da7176f7b0b6e87076527e0/unbored/activity.py#L45). Note too that the bindings are in effect if any of that widget's children are focused.

More generally, this is how bindings work in Textual. There's a slight aside I won't get into here, but in effect everything is a widget and a binding declared for a widget is good for that widget and all child widgets. Define for your app? That's good for all screens and widgets within. Defined for one screen? That's good for all widgets within. Defined for a particular widget? That's good for that widget and its children.

**Comment by jacobtomlinson** [comment]
Ok that makes sense. In that case I have two widgets that I want different (overlapping) bindings on. If neither of them is in focus I want one of them to be kept in focus. I.e none of the other widgets should be focusable. Do you have thoughts on a "correct" way to do that?

**Comment by davep** [comment]
What are the other widgets such that they can be focused but you don't want them to have focus? Normally I'd only use a focusable widget if I want it to have focus, otherwise I'd use something different that can't have focus.

**Comment by jacobtomlinson** [comment, marked as answer]
Heh I guess this is getting to the core of my misunderstanding. I assumed all widgets were always focusable (but I couldn't seem to get focus working either). From your question I'm guessing widget focus is opt in somehow which makes sense why I couldn't get my widget to focus. I can't figure out from the docs how to make a "focusable widget".

Maybe to rephrase my question. How do I make two of my widgets focusable? And how do I ensure that I can't be in a fully blurred state (one widget must always be in focus)?

_(Feel free to redirect me to the forum if that's a better place for this question)_

**Reply by davep** [reply]
Perhaps, at this point, it might be helpful for you to show a small self-contained example of what it is you're trying to do? I suspect I'd find it easier to suggest options if I knew what sort of widget you were trying to make.

More generally though, by default widgets (that's things that inherit from `Widget`) can't and don't receive focus. Their `can_focus` property has to be set to `True` for this to happen. So, if you look [at Textual's `Button` widget](https://github.com/Textualize/textual/blob/d2ba22b86f48f4ce5b0f55767efdcf1a5478b180/src/textual/widgets/_button.py#L25) for example, you'll see it inherits from `Static` (which also can't receive focus) and also sets `can_focus` to `True`.

As I say, a self-contained example of what you're trying to do would be helpful for me to understand how best to suggest options. That said, perhaps this little demo helps?

```python
from textual.app import App, ComposeResult
from textual.widgets import Label, Input, Button

class MyOwnSuperCoolFocusThing( Label, can_focus=True ):

    DEFAULT_CSS = """
    MyOwnSuperCoolFocusThing {
        border: solid darkred;
    }

    MyOwnSuperCoolFocusThing:focus {
        border: solid red;
    }
    """

class FocusDemo( App[ None ] ):

    CSS = """
    Screen {
        align: center middle;
    }
    """

    def compose(self) -> ComposeResult:
        yield Label( "This is a label, it can't get focus" )
        yield Input( placeholder="This is an input, it can get focus" )
        yield Button( "This is a button, it can get focus" )
        yield MyOwnSuperCoolFocusThing( "While this is a label, I've made it so it can focus" )

    def on_mount( self ) -> None:
        # When the app starts, we force focus to the input and then focus
        # won't be lost again.
        self.query_one( Input ).focus()

if __name__ == "__main__":
    FocusDemo().run()
```

**Reply by jacobtomlinson** [reply]
Ahh I think knowing about `can_focus` is all I need to get things working. I wonder if adding a tip to the input docs on making focusable widgets would be useful for folks in the future https://textual.textualize.io/guide/input/#input-focus. 

Much appreciated!

**Comment by davep** [comment]
> Feel free to redirect me to the forum if that's a better place for this question

Aye, I think this is turning more into a Q&A, I'll move this to there.
